radsat-sk/operation/subsystems/RPdb.c

```diff
@@ -0,0 +1,272 @@
+
+	i2c_received = PdbI2cTalk(PdbVoltageCommandBytes[1]);
+	dataStorage.outputVoltageBatteryBus = ADC_COUNT_TO_BATTERY_BUS_OUTPUT_VOLTAGE * i2c_received;
+
+	i2c_received = PdbI2cTalk(PdbVoltageCommandBytes[2]);
+	dataStorage.outputVoltage5VBus = ADC_COUNT_TO_5V_BUS_OUTPUT_VOLTAGE * i2c_received;
+
+	i2c_received = PdbI2cTalk(PdbVoltageCommandBytes[3]);
+	dataStorage.outputVoltage3V3Bus = ADC_COUNT_TO_3V3_BUS_OUTPUT_VOLTAGE * i2c_received;
+
+
+	// Get ADC Output current readings from the PDB
+	i2c_received = PdbI2cTalk(PdbCurrentCommandBytes[0]);
+	dataStorage.outputCurrentBCR_mA = ADC_COUNT_TO_BCR_OUTPUT_CURRENT * i2c_received;
+
+	i2c_received = PdbI2cTalk(PdbCurrentCommandBytes[1]);
+	dataStorage.outputCurrentBatteryBus = ADC_COUNT_TO_BATTERY_BUS_OUTPUT_CURRENT * i2c_received;
+
+	i2c_received = PdbI2cTalk(PdbCurrentCommandBytes[2]);
+	dataStorage.outputCurrent5VBus = ADC_COUNT_TO_5V_BUS_OUTPUT_CURRENT * i2c_received;
+
+	i2c_received = PdbI2cTalk(PdbCurrentCommandBytes[3]);
+	dataStorage.outputCurrent3V3Bus = ADC_COUNT_TO_3V3_BUS_OUTPUT_CURRENT * i2c_received;
+
+
+	// Get ADC temperature reading from the PDB
+	i2c_received = PdbI2cTalk(0x10E308);	// Telemetry code for motherboard temperature
+	dataStorage.PdbTemperature = (ADC_COUNT_TO_MOTHERBOARD_TEMP_SCALE * i2c_received)
+										- ADC_COUNT_TO_MOTHERBOARD_TEMP_SHIFT ;
+
+	return dataStorage;
+}
+
+
+/**
+ * Pet the Communications watchdog on the PDB using code 0x22
+ * @return either an error if it occurred, or 0
+ */
+int petPdbWatchdog(void){
+
+	// One way communication so just use transmit using reset watchdog command 0x22
+	int error = i2cTransmit(PDB_I2C_SLAVE_ADDR, 0x2200, PDB_COMMAND_LENGTH);
+	if (error != 0) {
+		return error;
+	}
+	else {
+		return 0;
+	}
+}
+
+
+
+/***************************************************************************************************
+                                         PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+
+static uint16_t PdbI2cTalk(uint32_t command) {
+
+	// Create a temporary variable for receiving I2C data
+	uint16_t i2c_data = 0;
+
+	// Create a temp variable for passing the proper command length
+	uint8_t comm_length;
+	uint8_t comm_delay;
+
+	// Variable selection for command length dependent on if it is a telemetry call
+	// Byte mask to look only at the command field to see if it is 0x10 (telem call)
+	if ((command && 0x0000FF0000) == 0x0000100000) {
+		comm_length = PDB_TELEM_COMMAND_LENGTH;
+		comm_delay = PDB_I2C_TELEM_DELAY;
+	}
+	else {
+		comm_length = PDB_COMMAND_LENGTH;
+		comm_delay = PDB_I2C_DELAY;
+	}
+
+	// tell PDB to give us data; and store into our internal buffer
+	int error = i2cTalk(PDB_I2C_SLAVE_ADDR, comm_length, PDB_RESPONSE_LENGTH,
+							command, i2c_data, comm_delay);
+
+	// return error? if an error occurs, else send the data back
+	// TODO: Make this return 0 on a failure? Printf the error instead? Get advice for this
```

Comment:
This will get cleared up when Ty's main cleanup gets merged with alpha because there's some useful examples